generalize mutation models

petrelharp opened the pull request:
Here's how I'd like to make mutation models more flexible, as outlined in #1006. Mutation models would only need to have the 'pick root allele' and 'transition allele' methods; what we've done so far is the special case of MutationMatrixModels. I've coded the (small) changes up in python, with the SLiM mutation model.

I've also made the check that we aren't sticking a mutation above an existing one optional (in python only, again).

How's this look?

Note: with this in place, it'll be easy to do models of microsats, etcetera.

jeromekelleher commented on tests/test_mutations.py:
```diff
@@ -27,11 +27,11 @@
 import _msprime
 import msprime
 import tests.wright_fisher as wf
+import tskit
```
tskit is third party - these are *local* imports.

petrelharp commented:
Huh - the pre-commit failed because for some reason I had a `tskit/` directory in my repo, with only a `__pycache__` in it, and so `pre-commit` thought that tskit was a local repo so wanted to reorder imports.

petrelharp commented:
> modulo the text processing, which will be tedious

Can we use `strcmp` and related tools?

jeromekelleher commented:
> Can/should we use strcmp and related tools?

Sure, whatever is simplest. We'll just have to look very hard at it to make sure that everything is correct. Memory management for variable length strings like this is painful.

petrelharp commented:
This is not yet working (segfaults), but input on the general approach would be good. Not sure at all this is the right way to go.

petrelharp commented:
Exciting update! This passes tests (er, CI notwithstanding??)!  I'll stop force-pushing.

So, the idea is that a mutation model has to define two method: `root_allele( )` and `transition_allele( )`. They each return ints (which is necessary so they can error). And, currently, they provide output by being passed in a `**char` which then gets its contents copied over into the correct place.  For the mutation matrix model case, wehre the possible alleles are fixed, this is natural. But, for the next (SLiM) case, each allele will be new, and so these methods will have to allocate some memory. We can just malloc, but then that's lots of little mallocs. Or, we could give the slim mutation model its own blkalloc?  Or, a totally different approach?

petrelharp commented on lib/msprime.h:
```diff
@@ -318,25 +318,25 @@ typedef struct demographic_event_t_t {
+    union {
+        mutation_matrix_t mutation_matrix;
+        /* Other known mutation models */
+    } params;
```
oh, nice!

petrelharp commented:
Oh, gee, that's much more elegant, good.

Thinking ahead to the SLiM version, I wonder if we want to do some of this slightly differently. Right now, `transition` gives a pointer to the allele's pre-existing location in memory, which gets copied into `mut->derived_state`. For SLiM mutations, we'll have to create the allele somewhere just for it to be immediately copied over to `mut->derived_state`. This is an extra alloc and copy that would be unnecessary if `transition` was in charge of setting `mut->derived_state` directly. The main advantage to the current set-up is that `transition` won't need to alloc... but, for SLiM mutations it'll have to alloc anyhow.  So, I'm thinking that we should make `transition` (and `choose_root`) in charge of allocating memory too, maybe by adding a `blkalloc` attribute to `mutation_model_t`?  Or by sharing with mutgen, somehow?

A related issue is that for the SLiM generator we'll want to add metadata to the new mutations also. So, maybe it makes sense for `transition` to just get a pointer to the mutation object, and `choose_root` a pointer to the site object, and that they are responsible for updating appropriately (including setting `keep`, for mutations).

I'm happy to have a go at implementing this, but some input would be very helpful first.

jeromekelleher commented:
That's a good idea @petrelharp - maybe the two key functions should get the site_t and mutation_t objects as parameters, which they are responsible for updating. The ``mutation_model_t`` could have a blkalloc that's for doing the various mallocs needed for the Site and Mutation information.

petrelharp commented:
Ok - I've done a refactor here, so that now the signatures are
```
int mutation_model_choose_root_state(mutation_model_t *self,
        gsl_rng *rng, tsk_site_t *site);
int mutation_model_transition(mutation_model_t *model, gsl_rng *rng,
        const char *parent_allele, tsk_size_t parent_allele_length,
        tsk_mutation_t *mutation);
```
Two things:

1.  `copy_string` doesn't work because I don't understand `blkalloc`: replacing the `blkalloc` with a `malloc` fixes it; something to do with the block size? Help?

2. Previously, we had mutgen's `mutation_t` mostly like `tsk_mutation_t` but with some extra stuff (eg `new` and `keep`). But then we had to pass in a `mutation_t` to `mutation_model_transition`, and so `mutation_t` had to be public, which seemed confusing. So, I changed `mutation_t` to carry a pointer to a `tsk_mutation_t`, which is the thing passed in to `mutation_model_transition`. This means that `mutation_model_transition` can't set the `keep` parameter directly, but this seems OK.

jeromekelleher commented:
It's not immediately clear to me what the problem with the block_allocator is @petrelharp, but I'm having a hard time getting things straight in my head.

I think it would be a lot simpler if we just forget about the existence of tsk_site_t and tsk_mutation_t entirely. I see no problem in defining ``msp_site_t`` and ``msp_mutation_t`` in ``msprime.h``. Sure, they're quite like the tsk_ counterparts, but so what? We're not going to be changing ``tsk_site_t``, e.g., ever if we can help it so I don't think it's a maintenance burden having two similar types defined across the different projects. 

I can go ahead and make these changes and (hopefully) simplify the memory management if you like?

petrelharp commented:
Ok! That'd be great. Go for it.

petrelharp commented:
Well:mMy concern was that the extra stuff in `mutation_t` doesn't feel like we want to commit it to the public API, as opposed to `tsk_mutation_t`.

jeromekelleher commented:
> Well: my concern was that the extra stuff in mutation_t doesn't feel like we want to commit it to the public API, as opposed to tsk_mutation_t.

The msprime C API isn't public anyway, we can do whatever we want.

jeromekelleher commented:
I've tidied things up @petrelharp and made it work again. The memory management is much simpler now, and I think the algorithm flow is more straightforward. It has turned out to be quite time consuming, though.

However, I'll have to rewrite the Python C API side of things completely to make the model polymorphism work, and unfortunately I don't have time for this. I'm not quite sure how this will work, so it'd be better to leave it with me. I'll try and get some time to implement the SLiM model in C and update here when I get a chance.

petrelharp commented:
Ah, ok, that's simpler, and more memory efficient! Nice.

How about I take a first pass at the SLiM model? It's pretty clear in my head what it should do, and if I malloc instead of blkalloc'ing, then I'm sure I can make it work.

jeromekelleher commented:
That'd be really helpful, thanks. Good test coverage on the C code would be super helpful too. I can wrangle the top-to-bottom structure then so we also work in Python.

petrelharp commented:
Ok - here's a go at the SLiM mutator. Note that I've had to also pass in the parental metadata, since we need to append to that.

FYI, the model here is that each new mutation gets a unique ID (assigned sequentially starting from a number that's passed in), which is appended (as bytes) to the previous allele to make the derived state; and, a corresponding metadata entry (which just describes the mutation type, and so will always be the same) is appended to the metadata.  There's a pretty good C test that's checking that's exactly what's happening.

jeromekelleher commented on lib/mutgen.c:
```diff
@@ -292,6 +294,101 @@ mutation_matrix_free(mutation_model_t *self)
+
+/* Copy of MutationMetadataRec in slim_sim.h: */
+typedef struct __attribute__((__packed__)) {
+    int32_t mutation_type_id_; // 4 bytes (int32_t): the id of the mutation type the mutation belongs to
```
We can get rid of the trailing underscores here I'd say, might as well use the local coding conventions. A bit more details on the provenance would be help here I think - what date/what git hash were these structs copied at?

petrelharp replied:
I'll put in more details to the reference. But, I guess I vote to keep them exactly matching, as a reminder that we're copying this from elsewhere (so it's not to be changed).

jeromekelleher replied:
OK, that sounds good.

jeromekelleher commented:
Looks good @petrelharp. We'll need to change to useing a blkalloc though, as it's currently leaking memory.

petrelharp commented:
> Looks good @petrelharp. We'll need to change to useing a blkalloc though, as it's currently leaking memory.

I did?

jeromekelleher commented on lib/mutgen.c:
```diff
@@ -187,93 +184,348 @@ mutation_model_check_validity(mutation_model_t *self)
+    memset(self, 0, sizeof(mutation_model_t));
+    
+    /* In practice this is the minimum we can support */
+    block_size = GSL_MAX(block_size, 128);
```
Where does the 128 come from here? That seem very small to me. Unless you've got a good reason, I would say the best behaviour here is to have a default of say 4K, which we use if the input block size is 0. We don't ever set these in practise, so I'm not sure there's much point in having the parameter.

petrelharp replied:
er, wherever I copied the blkalloc from? sorry, I forgot to track down what this means (and what it should be)

jeromekelleher replied:
Just put in a fixed size of 4K here. There's no point in having this knob to tune, it only makes a difference for really big stuff.

petrelharp replied:
I've set it to 8192, following what's in `mutgen_alloc` and `mutgen_init_allocator`.

jeromekelleher commented on lib/mutgen.c:
```diff
@@ -187,93 +184,348 @@ mutation_model_check_validity(mutation_model_t *self)
+        goto out;
+    }
+
+    slim_md = malloc(sizeof(*slim_md));
```
I'm not sure there's much point in storing this in the struct - since it's a fixed size, it would be simpler/more obvious what's going on if we alloced it statically in the function that's using it.

jeromekelleher commented:
I'm not sure why this isn't getting full CI checks  - maybe worth a rebase here @petrelharp?

petrelharp commented:
Ok, I've removed the slim_metadata property and put it in the transition method; and also rebased, doing a bit of squashing as well.  I'm still getting seemingly unrelated errors from the local tests, strangely.

petrelharp commented:
The windows build was failing because Windows doesn't do packed; there are some other possible fixes [here](https://stackoverflow.com/questions/1537964/visual-c-equivalent-of-gccs-attribute-packed), but I've just implemented a lower-tech solution.

petrelharp commented:
I'm going to leave this alone now - want to have another look, @jeromekelleher? (and, thanks!!)

petrelharp commented:
![Screenshot from 2020-06-08 12-15-18](https://user-images.githubusercontent.com/1046249/84070915-bb82ac80-a981-11ea-934c-79069a877098.png)
<relaxed sigh>

jeromekelleher commented on tests/test_mutations.py:
```diff
@@ -1087,6 +1087,60 @@ def test_arbitrary_model(self):
+        for s in ts.sites():
+            self.assertEqual(s.ancestral_state, "")
+            for m in s.mutations:
+                a = tuple(m.derived_state.split(","))
```
This is different to the C generator now though, isn't it? There we're using little-endian 8 byte values and here we're using comma delimited text. Which one does SLiM do?

petrelharp replied:
Whoops, yes. SLiM does the [concatenated bytes thing](https://github.com/MesserLab/SLiM/blob/1c85d00a0417416431ca69f2cb68a1bcc134ec3d/core/slim_sim.cpp#L5137), not text. I was waiting to compare these until we've hooked the slim generator up to python.

jeromekelleher commented:
Looks good to me @petrelharp. The low-tech solution to packing bytes looks just right. I am confused about the exact format here, as we seem to be doing different things in the C and Python versions. But, I guess the next step is to setup the linkage between C and Python. Or, we can tidy this much up and merge first, I don't mind.

petrelharp commented:
> I guess the next step is to setup the linkage between C and Python. 

It'd be nice to do this before merging so we can properly test it (including I'll check if SLiM can read the mutations we've produced). Before you said you wanted to set up the linkage; is that still true?

jeromekelleher commented:
> It'd be nice to do this before merging so we can properly test it (including I'll check if SLiM can read the mutations we've produced). Before you said you wanted to set up the linkage; is that still true?

I can do this, unless you're super keen. Will I jump in now?

jeromekelleher commented:
How about I do some coding here if you do some reviewing on #1025 @petrelharp?

jeromekelleher commented:
I've added the basic plumbing here @petrelharp, should all basically work now. Good bit of work left in tidying everything up and making tests, but you should be able to test it against SLiM now anyway.

petrelharp commented on _msprimemodule.c:
```diff
@@ -4682,13 +4838,23 @@ PyInit__msprime(void)
-    PyModule_AddObject(module, "MutationModel", (PyObject *) &MutationModelType);
+    Py_INCREF(&SlimMutationModelType);
+    PyModule_AddObject(module, "SlimMutationModel",
+            (PyObject *) &SlimMutationModelType);
```
whoo, yep, I'm glad you did this bit!

petrelharp commented on lib/mutgen.c:
```diff
@@ -491,31 +490,35 @@ mutation_matrix_model_alloc(mutation_model_t *self, size_t num_alleles, char **a
 
 int MSP_WARN_UNUSED
-slim_mutator_alloc(mutation_model_t *self, int32_t mutation_type_id,
+slim_mutation_model_alloc(mutation_model_t *self, int32_t mutation_type_id,
```
good resolution to the inconsistent naming problem!

petrelharp commented:
Gah! I've just been around in about five circles until I rediscovered/remembered that (a) SLiM keeps track of derived states as bytes in its *internal tables*, but since these have to be ascii for tskit, converts them to comma-separated strings on writing them out.

jeromekelleher commented:
> Gah! I've just been around in about five circles until I rediscovered/remembered that (a) SLiM keeps track of derived states as bytes in its internal tables, but since these have to be ascii for tskit, converts them to comma-separated strings on writing them out.

Ah, well that's a relief in a way. I was getting worried that we'd have to do some hacks to support arbitrary bytes in the ancestral/derived states.

petrelharp commented:
> Ah, well that's a relief in a way. I was getting worried that we'd have to do some hacks to support arbitrary bytes in the ancestral/derived states.

Our former selves decided not to do this, I think.

jeromekelleher commented:
i've updated this @petrelharp, and it should be nearly ready now. I'll add an InfiniteSitesModel too though, before merging.

jeromekelleher commented:
This is ready for merging I think @petrelharp - can you have a quick look over please? There's still some loose ends re the API, but these can be tidied up with #1070.

petrelharp commented on _msprimemodule.c:
```diff
@@ -2112,18 +2314,41 @@ MutationGenerator_init(MutationGenerator *self, PyObject *args, PyObject *kwds)
+    } else {
+        PyErr_SetString(PyExc_TypeError,
+            "model must be an instance of MatrixMutationModel "
+            "or SlimMutationModel");
```
or InfiniteAllelesMutationModel

petrelharp commented on lib/mutgen.c:
```diff
@@ -187,93 +184,494 @@ mutation_model_check_validity(mutation_model_t *self)
+    /* We allow for a potential comma to separate previous elements in the
+     * list. We allow for a possible comma to separate the previous element
+     * in the list, as well as a NULL byte added by snprintf. We don't bother
+     * trying to alloc the exact number of bytes needed. */
```
duplicated sentence in the comment. (also, good call here)

petrelharp commented on lib/mutgen.c:
```diff
@@ -187,93 +184,494 @@ mutation_model_check_validity(mutation_model_t *self)
+
+/* The maximum number of digits for an unsigned 64 bit integer is 20
+ * and one byte for the NULL terminator. */
+#define MAX_UINT_BUFF_SIZE 21
```
we could use this above (`max_digits` in slim_transition_allele) as well

jeromekelleher replied:
I thought about this, but it's a bit awkward as we'd have to have another implementation of make_allele as we're using uint64 in one and int64 in the other. The reason for having ``uint64`` here, btw, is so that we have well defined overflow semantics. Unsigned ints are guaranteed to wrap around to zero, whereas signed ints are undefined. Having overflow seemed simpler than having to test for it, but we could totally do it same way as the slim model too.

Anyway, it didn't seem worth complicating the code more, as the slim code is pretty water tight and reasonably simple now. The extra few bytes spent per allele is trivial.

petrelharp commented on lib/mutgen.c:
```diff
@@ -187,93 +184,494 @@ mutation_model_check_validity(mutation_model_t *self)
+    memcpy(buff, tmp_buff, len);
+    params->next_allele++;
+    *dest = buff;
+    *dest_length = (tsk_size_t) len;
```
`len` was already cast to `tsk_size_t` on assignment - but, why not just use `len` instead of `err` throughout, as in `slim_transition_allele`?

jeromekelleher replied:
Good spot on the cast. We're passing ``len`` directly to blk_alloc_get as well as assigning it, so we'd have to cast it twice if we used the return value from sprintf. I thought this was the simplest and clearest. I've changed ``err`` to ``num_digits`` now though, as err was a bit misleading.

petrelharp commented on lib/mutgen.c:
```diff
@@ -483,81 +880,104 @@ mutgen_add_site(mutgen_t *self, double position, bool new, const char *ancestral
-    const char *metadata, tsk_size_t metadata_length)
+static int
+mutgen_add_mutation(
+    mutgen_t *self, site_t *site, node_id_t node, double time, mutation_t **new_mutation)
```
good call

petrelharp commented:
Looks great! Awesome!  A few minor suggestions, but it looks good!  I should check that things can be loaded correctly into SLiM - will report back.

jeromekelleher commented:
Great! I'm going to merge this now @petrelharp - we can follow up with updates if we need them, but let's get this big diff merged.
